### ISSUE TITLE
Slurm: Silent stale workers

### ISSUE THREAD

**Issue by pat-s** [mschubert/clustermq#181]
slurm/clustermq settings are the same as in #180 

Some workers are starting fine and do not throw an exception.
However, they are just silently stale.

This means, if I let all functioning workers run long enough to finish, I'd wait infinitely for some workers.

# How do I notice that?

When processing targets via _drake_ (cc @wlandau), each target is a assigned to a slurm worker.
Some targets process fine. Others error directly (randomly it seems) due to #180.

After 2 days or so, I still have plenty workers "running" Inspecting the logs, it turns out they are all stale. (Again, this does not apply to targets using a specific function or package.)
I can verify that they are stale by seeing that they started processing 2 days ago and are still at "iteration 1" whereas they should be finished by then.
(One iteration does not more than 5 mins in this case).

If I terminate all of them and restart the whole process, the same will happen: Some of the stale ones will process fine this time, others show the same behavior.

I am able to process all at some point by restarting the whole process 5 - 10 times, every time with less "targets" that are left to process.
The number of initial workers does not have an influence here, it will just scale up/down with the number of workers spawned.

---

So in summary it looks as follows for me right now. I run a workflow using _drake_ with (let's fix it at 100 to simplify it) workers without any R parallelization going on.

- ~ 30-50% of them error with #180 
- ~ 30-50% of them become stale right from the start
- ~ 30-50% of them process fine

Terminating and restarting everything 5-10 times I am able to get all targets processed. However, this is a very annoying process :/

All of this reads a bit "random", I know. I am happy about any suggestion what I could try to make the situation better.

**Comment by mschubert** [comment]
Unfortunately, we don't currently re-assign lost targets. In your example, if a worker crashes due to #180, the target is lost. This is also why your main session does not complete: the master is waiting for this target it never receives.

This is a major shortcoming and [will be addressed in v0.9](https://github.com/mschubert/clustermq/projects/5).

For your

> ~ 30-50% of them become stale right from the start

workers, do they actually finish their job or not? (i.e., what's in the worker log file using `%a`?)

**Comment by pat-s** [comment]
> Unfortunately, we don't currently re-assign lost targets. In your example, if a worker crashes due to #180, the target is lost. This is also why your main session does not complete: the master is waiting for this target it never receives.
> 
> This is a major shortcoming and will be addressed in v0.9.

Great 👍 

> do they actually finish their job or not? (i.e., what's in the worker log file using %a?)

No, they do not.
And this is also the major pain here.
I would wait for those infinitely if I would not look at the lock file and see that nothing is going on.

In these cases I am wondering even more since the workers are "fresh", meaning they are building their first target after the whole job was started and distributed across workers.

I know this is a tedious one and I cannot really come up with a reprex here.
However, due to the fact that I do not even get an error but just a "silent death" in the background this let's me worry a lot.
In my case, I can easily monitor and shut down workers/processes on the HPC but porting this to an external cluster would make this issue a huge pain (because re-queuing everything X times is not nice if the queue is long :/ ).

I have to inspect whether and how https://github.com/mschubert/clustermq/issues/180#issuecomment-568915117 has an influence on this issue here as well and will report back.

I'll close here for now since I cannot come up with something substantial.
